src/Riok.Mapperly/Descriptors/Enumerables/CollectionInfoBuilder.cs

```diff
@@ -188,21 +188,147 @@ or CollectionType.ReadOnlyMemory
+                    | CollectionType.IReadOnlySet
+                    | CollectionType.ISet
+                    | CollectionType.ICollection
+                    | CollectionType.IReadOnlyCollection
+                    | CollectionType.IEnumerable,
+            CollectionType.IImmutableSet => CollectionType.IImmutableSet | CollectionType.IReadOnlyCollection | CollectionType.IEnumerable,
+            CollectionType.ImmutableSortedSet
+                => CollectionType.ImmutableSortedSet
+                    | CollectionType.IImmutableSet
+                    | CollectionType.IList
+                    | CollectionType.IReadOnlyList
+                    | CollectionType.ISet
+                    | CollectionType.IReadOnlySet
+                    | CollectionType.ICollection
+                    | CollectionType.IReadOnlyCollection
+                    | CollectionType.IEnumerable,
+            CollectionType.ImmutableQueue => CollectionType.ImmutableQueue | CollectionType.IImmutableQueue | CollectionType.IEnumerable,
+            CollectionType.IImmutableQueue => CollectionType.IImmutableQueue | CollectionType.IEnumerable,
+            CollectionType.ImmutableStack => CollectionType.ImmutableStack | CollectionType.IImmutableStack | CollectionType.IEnumerable,
+            CollectionType.IImmutableStack => CollectionType.IImmutableStack | CollectionType.IEnumerable,
+            CollectionType.ImmutableDictionary
+                => CollectionType.ImmutableDictionary
+                    | CollectionType.IImmutableDictionary
+                    | CollectionType.IDictionary
+                    | CollectionType.IReadOnlyDictionary
+                    | CollectionType.ICollection
+                    | CollectionType.IReadOnlyCollection
+                    | CollectionType.IEnumerable,
+            CollectionType.IImmutableDictionary
+                => CollectionType.IImmutableDictionary
+                    | CollectionType.IReadOnlyDictionary
+                    | CollectionType.IReadOnlyCollection
+                    | CollectionType.IEnumerable,
+            CollectionType.ImmutableSortedDictionary
+                => CollectionType.ImmutableSortedDictionary
+                    | CollectionType.IImmutableDictionary
+                    | CollectionType.IReadOnlyDictionary
+                    | CollectionType.IReadOnlyCollection
+                    | CollectionType.IEnumerable,
+            CollectionType.Span => CollectionType.Span,
+            CollectionType.ReadOnlySpan => CollectionType.ReadOnlySpan,
+            CollectionType.Memory => CollectionType.Memory,
+            CollectionType.ReadOnlyMemory => CollectionType.ReadOnlyMemory,
+
+            CollectionType.None when SymbolEqualityComparer.Default.Equals(type, types.Get<string>()) => CollectionType.IEnumerable,
+            _ => IterateHierarchy(type, types)
+        };
+
+        static CollectionType IterateHierarchy(ITypeSymbol type, WellKnownTypes types)
```

Comment:
This code base usually explicitly states the visibility. I'd set this to private.